analyze: CFG_build: Handle labels and gotos

IagoAbal opened the pull request.

IagoAbal committed:
analyze: CFG_build: Handle labels and gotos

aryx commented on lang_GENERIC/analyze/CFG_build.ml:
```diff
@@ -64,6 +71,19 @@ let add_arc_opt (starti_opt, nodei) g =
+  !(state.gotos)
+  |> List.iter (fun (srci, label) ->
+    match Hashtbl.find_opt state.labels label with
+    | None -> Common.(pr2 @@ spf "Could not resolve label: %s" (fst label))
```
Can you avoid those Module.(xxx) because the ocaml parser in pfff does not understand them yet, and we use semgrep-on-ocaml in the CI of pfff and semgrep, which could prevent this PR to be accepted.

IagoAbal replied:
Sure!

aryx commented on lang_GENERIC/analyze/CFG_build.ml:
```diff
@@ -85,14 +105,17 @@ let add_arc_opt (starti_opt, nodei) g =
 *)
 
 
-let rec (cfg_stmt: state -> F.nodei option -> stmt -> F.nodei option) =
```
maybe a small comment here explaining this new return type.

aryx replied:
See the comment in the code above cfg_stmt that is not anymore up to date given your change.

aryx replied:
you can add a ver3: explaining your change.

aryx replied:
You can add a comment also about alternatives:
(* An alternative would be to do a first pass to associate a node for each label, and then a second pass
 * that uses this association *)

aryx commented on lang_GENERIC/analyze/CFG_build.ml:
```diff
@@ -34,6 +34,8 @@ module F = IL (* to be even more similar to controlflow_build.ml *)
 (* Types *)
 (*****************************************************************************)
 
+type cfg_label = string * G.sid
```
maybe a comment like (* like IL.label but without the token attached to the ident *)

IagoAbal replied:
Done!

aryx reviewed (approved):
Very nice!
It's nice that your solution does not need a new NLabel node and you can do it in one pass.
I would just like a few more comments here and there to explain the type
(like reading cfg_stmt_list help with the `Cfg (firsti, lasti) names. Would be good to explain that a bit earlier.
Maybe it's better to introduce a new type like type cfg_stmt_result = Label of ... | CfgFirstLast where you could explain a bit things.

IagoAbal commented:
Thanks for the feedback! I agree the code was lacking some documentation. I have added some comments to the code, also did some small cleanup, and fixed a small regression (there should not be a "last" node for `Return` since it exits the function).

By the way, I was considering whether to take the `gotos` pending list out of the `state`. It is not strictly needed, they could be accumulated in `cfg_stmt_list` in a pure list. But if I do that, I either add an extra `Goto` constructor to `cfg_stmt_result` or an extra field to `CfgFirstLast`... I am a bit undecided so I kept it as it is for now. If you have a preference please let me know.

aryx reviewed (approved):
looks good!
I'm merging it.
